state/pruner_test.go

```diff
@@ -0,0 +1,181 @@
+
+	err := pruner.SetIndexerRetainHeight(int64(4))
+	require.NoError(t, err)
+
+	actual, err := pruner.GetIndexerRetainHeight()
+	require.NoError(t, err)
+
+	require.Equal(t, int64(4), actual)
+
+	pruner.PruneIndexesToRetainHeight(4)
+
+	keysAfterPrune4 := setDiff(kv.GetKeys(txIndexer), metaKeys)
+	require.Equal(t, keysAfterPrune4, setDiff(keys[3], keys[2]))
+
+	events, txResult1, txResult2 := getEventsAndResults(1)
+	//publish block with events
+	err = eventBus.PublishEventNewBlockEvents(events)
+	require.NoError(t, err)
+
+	err = eventBus.PublishEventTx(types.EventDataTx{TxResult: *txResult1})
+	require.NoError(t, err)
+
+	err = eventBus.PublishEventTx(types.EventDataTx{TxResult: *txResult2})
+	require.NoError(t, err)
+
+	time.Sleep(100 * time.Millisecond)
+
+	keys14 := kv.GetKeys(txIndexer)
+
+	pruner.PruneIndexesToRetainHeight(4)
+	keysSecondPrune4 := kv.GetKeys(txIndexer)
+
+	require.Equal(t, keys14, keysSecondPrune4)
+}
+
+func createTestSetup(t *testing.T) (*sm.Pruner, *kv.TxIndex, indexer.BlockIndexer, *types.EventBus) {
+	config := test.ResetTestRoot("pruner_test")
+	// event bus
+	eventBus := types.NewEventBus()
+	eventBus.SetLogger(log.TestingLogger())
+	err := eventBus.Start()
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		if err := eventBus.Stop(); err != nil {
+			t.Error(err)
+		}
+		err := os.RemoveAll(config.RootDir)
+		if err != nil {
+			t.Error(err)
+		}
+	})
+
+	// tx indexer
+	memDB := db.NewMemDB()
+	txIndexer := kv.NewTxIndex(memDB)
+	blockIndexer := blockidxkv.New(db.NewPrefixDB(memDB, []byte("block_events")))
+
+	blockDB := db.NewMemDB()
+	stateDB := db.NewMemDB()
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: false,
+	})
+	bs := store.NewBlockStore(blockDB)
+	pruner := sm.NewPruner(stateStore, bs, blockIndexer, txIndexer, log.TestingLogger())
+
+	return pruner, txIndexer, blockIndexer, eventBus
+}
+
+func getEventsAndResults(height int64) (types.EventDataNewBlockEvents, *abci.TxResult, *abci.TxResult) {
+	events := types.EventDataNewBlockEvents{
+		Height: height,
+		Events: []abci.Event{
+			{
+				Type: "begin_event",
+				Attributes: []abci.EventAttribute{
+					{
+						Key:   "proposer",
+						Value: "FCAA001",
+						Index: true,
+					},
+				},
+			},
+		},
+		NumTxs: int64(2),
+	}
+	txResult1 := &abci.TxResult{
+		Height: height,
+		Index:  uint32(0),
+		Tx:     types.Tx(fmt.Sprintf("foo%d", height)),
+		Result: abci.ExecTxResult{Code: 0},
+	}
+	txResult2 := &abci.TxResult{
+		Height: height,
+		Index:  uint32(1),
+		Tx:     types.Tx(fmt.Sprintf("bar%d", height)),
+		Result: abci.ExecTxResult{Code: 0},
+	}
+	return events, txResult1, txResult2
+}
+
+func isSubset(smaller [][]byte, bigger [][]byte) bool {
+	for _, elem := range smaller {
+		if !slices.ContainsFunc(bigger, func(i []byte) bool {
+			return bytes.Equal(i, elem)
+		}) {
+			return false
+		}
+	}
+	return true
+}
+
+func isEqualSets(x [][]byte, y [][]byte) bool {
+	return isSubset(x, y) && isSubset(y, x)
```

Comment:
You can check first whether the lengths are equal, to avoid scanning if they are not.

Reply:
I don't really want to do that, cause formally speaking there might be duplicates, so two slices of different lengths might be equal as sets. 
Also, those slices in all cases are of size < 100, so that's a bit of an overkill optimization... 
And that's only used in testing:) 

But if you want to see that, I would do, not a big deal:)

Reply:
Ah so you can have duplicates and then you consider them as equal. Got it, that is fine then. Also as you said, if its just for testing the perf gain is not that big of a deal.